File: vin-decoder-extract/ts/services/scan-history.service.ts
// ============================================================
// BYKI WORKSHOP DIAGNOSTIC - SCAN HISTORY SERVICE
// Firebase/Firestore CRUD for vehicle scan sessions
// ============================================================

import { db, isConfigured } from '@/lib/firebase/config';
import {
  collection,
  addDoc,
  query,
  where,
  orderBy,
  limit,
  getDocs,
  serverTimestamp,
} from 'firebase/firestore';
import type {
  VehicleScan,
  ECUInfo,
  DiagnosticTroubleCode,
  LiveDataParameter,
  MonitoringTestResult,
  ReadinessMonitors,
  DiagnosticCase,
  ClearResults,
  FreezeFrameData,
} from '@/types';

// ============ Types ============

export interface SaveScanData {
  vin: string;
  brand: string;
  model: string;
  year: number;
  ecusFound: ECUInfo[];
  confirmedDtcs: DiagnosticTroubleCode[];
  pendingDtcs: DiagnosticTroubleCode[];
  permanentDtcs: DiagnosticTroubleCode[];
  freezeFrames: Record<string, FreezeFrameData>;
  monitoringTests: MonitoringTestResult[];
  readinessMonitors: ReadinessMonitors | null;
  liveDataSnapshot: LiveDataParameter[];
  dtcsCleared: ClearResults | null;
  diagnosticCases: DiagnosticCase[];
  overallHealthScore: number;
}

// ============ Service ============

class ScanHistoryService {
  private collectionName = 'vehicleScans';

  /**
   * Save a complete scan session to Firestore
   */
  async saveScan(data: SaveScanData): Promise<string | null> {
    if (!isConfigured || !db) {
      console.warn('Firebase not configured — scan not saved');
      return null;
    }

    try {
      // Serialize DTC objects to plain data for Firestore
      const docData = {
        vin: data.vin,
        scannedAt: serverTimestamp(),
        brand: data.brand,
        model: data.model,
        year: data.year,
        ecusFound: data.ecusFound.map((e) => ({
          type: e.type,
          name: e.name,
          address: e.address,
          status: e.status,
        })),
        confirmedDtcs: data.confirmedDtcs.map((d) => ({
          code: d.code,
          description: d.definition.description,
          severity: d.definition.severity,
        })),
        pendingDtcs: data.pendingDtcs.map((d) => ({
          code: d.code,
          description: d.definition.description,
        })),
        permanentDtcs: data.permanentDtcs.map((d) => ({
          code: d.code,
          description: d.definition.description,
        })),
        freezeFrames: Object.entries(data.freezeFrames).reduce(
          (acc, [code, ff]) => ({
            ...acc,
            [code]: {
              engineRPM: ff.engineRPM,
              vehicleSpeed: ff.vehicleSpeed,
              coolantTemp: ff.coolantTemp,
              engineLoad: ff.engineLoad,
            },
          }),
          {}
        ),
        monitoringTests: data.monitoringTests.map((t) => ({
          component: t.component,
          healthPct: t.healthPct,
          passed: t.passed,
        })),
        readinessMonitors: data.readinessMonitors
          ? {
              milOn: data.readinessMonitors.milOn,
              monitors: data.readinessMonitors.monitors.map((m) => ({
                name: m.name,
                available: m.available,
                complete: m.complete,
              })),
            }
          : null,
        overallHealthScore: data.overallHealthScore,
        dtcCount:
          data.confirmedDtcs.length +
          data.pendingDtcs.length +
          data.permanentDtcs.length,
      };

      const docRef = await addDoc(collection(db, this.collectionName), docData);
      return docRef.id;
    } catch (error) {
      console.error('Failed to save scan:', error);
      return null;
    }
  }

  /**
   * Load scan history for a specific VIN (most recent first)
   */
  async getHistoryByVin(vin: string, maxResults = 10): Promise<VehicleScan[]> {
    if (!isConfigured || !db) {
      return [];
    }

    try {
      const q = query(
        collection(db, this.collectionName),
        where('vin', '==', vin),
        orderBy('scannedAt', 'desc'),
        limit(maxResults)
      );

      const snapshot = await getDocs(q);
      return snapshot.docs.map((doc) => {
        const data = doc.data();
        return {
          id: doc.id,
          vin: data.vin,
          scannedAt: data.scannedAt?.toDate?.()?.toISOString() || new Date().toISOString(),
          brand: data.brand || '',
          model: data.model || '',
          year: data.year || 0,
          confirmedDtcs: (data.confirmedDtcs || []).map((d: any) => d.code),
          pendingDtcs: (data.pendingDtcs || []).map((d: any) => d.code),
          permanentDtcs: (data.permanentDtcs || []).map((d: any) => d.code),
          healthScore: data.overallHealthScore ?? null,
          monitoringTests: data.monitoringTests || [],
        };
      });
    } catch (error) {
      console.error('Failed to load scan history:', error);
      return [];
    }
  }
}

export const scanHistoryService = new ScanHistoryService();

File: vin-decoder-extract/dart/lib/services/scan_history_service.dart
/// Scan history service — ported from scan-history.service.ts
/// Persists completed diagnostic scans to Firestore for
/// vehicle history tracking and trend analysis.

import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/diagnostic_types.dart';

/// Data needed to save a completed scan.
class SaveScanData {
  final String vin;
  final String brand;
  final String model;
  final int year;
  final List<ECUInfo> ecusFound;
  final List<DiagnosticTroubleCode> confirmedDtcs;
  final List<DiagnosticTroubleCode> pendingDtcs;
  final List<DiagnosticTroubleCode> permanentDtcs;
  final Map<String, FreezeFrameData> freezeFrames;
  final List<MonitoringTestResult> monitoringTests;
  final ReadinessMonitors? readinessMonitors;
  final List<LiveDataParameter> liveDataSnapshot;
  final ClearResults? dtcsCleared;
  final List<DiagnosticCase> diagnosticCases;
  final int overallHealthScore;

  const SaveScanData({
    required this.vin,
    required this.brand,
    required this.model,
    required this.year,
    required this.ecusFound,
    required this.confirmedDtcs,
    required this.pendingDtcs,
    required this.permanentDtcs,
    this.freezeFrames = const {},
    this.monitoringTests = const [],
    this.readinessMonitors,
    this.liveDataSnapshot = const [],
    this.dtcsCleared,
    this.diagnosticCases = const [],
    required this.overallHealthScore,
  });
}

class ScanHistoryService {
  static const _collectionName = 'vehicleScans';

  FirebaseFirestore get _firestore => FirebaseFirestore.instance;

  /// Save a completed scan to Firestore. Returns document ID or null on error.
  Future<String?> saveScan(SaveScanData data) async {
    try {
      final doc = await _firestore.collection(_collectionName).add({
        'vin': data.vin,
        'brand': data.brand,
        'model': data.model,
        'year': data.year,
        'scannedAt': FieldValue.serverTimestamp(),
        'ecusFound': data.ecusFound.map((e) => {
          'id': e.id,
          'name': e.name,
          'type': e.type.name,
          'address': e.address,
          'status': e.status.name,
        }).toList(),
        'confirmedDtcs': data.confirmedDtcs.map((d) => {
          'code': d.code,
          'description': d.description,
          'severity': d.severity?.name,
        }).toList(),
        'pendingDtcs': data.pendingDtcs.map((d) => {
          'code': d.code,
          'description': d.description,
          'severity': d.severity?.name,
        }).toList(),
        'permanentDtcs': data.permanentDtcs.map((d) => {
          'code': d.code,
          'description': d.description,
          'severity': d.severity?.name,
        }).toList(),
        'diagnosticCaseCount': data.diagnosticCases.length,
        'overallHealthScore': data.overallHealthScore,
        'totalDtcCount': data.confirmedDtcs.length +
            data.pendingDtcs.length +
            data.permanentDtcs.length,
      });
      return doc.id;
    } catch (e) {
      // Silently fail — scan history is non-critical
      return null;
    }
  }

  /// Retrieve scan history for a given VIN.
  Future<List<VehicleScan>> getHistoryByVin(String vin, {int maxResults = 10}) async {
    try {
      final snapshot = await _firestore
          .collection(_collectionName)
          .where('vin', isEqualTo: vin)
          .orderBy('scannedAt', descending: true)
          .limit(maxResults)
          .get();

      return snapshot.docs.map((doc) {
        final d = doc.data();
        return VehicleScan(
          id: doc.id,
          vin: d['vin'] ?? '',
          brand: d['brand'] ?? '',
          model: d['model'] ?? '',
          year: d['year'] ?? 0,
          scannedAt: ((d['scannedAt'] as Timestamp?)?.toDate() ?? DateTime.now()).toIso8601String(),
          totalDtcCount: d['totalDtcCount'] ?? 0,
          overallHealthScore: d['overallHealthScore'] ?? 0,
        );
      }).toList();
    } catch (e) {
      return [];
    }
  }
}

/// Singleton instance.
final scanHistoryService = ScanHistoryService();
